File: conformance/tests/function-patterns/function-patterns.runar.zig
const runar = @import("runar");

pub const FunctionPatterns = struct {
    pub const Contract = runar.StatefulSmartContract;

    owner: runar.PubKey,
    balance: i64 = 0,

    pub fn init(owner: runar.PubKey, balance: i64) FunctionPatterns {
        return .{ .owner = owner, .balance = balance };
    }

    fn requireOwner(self: *const FunctionPatterns, sig: runar.Sig) void {
        runar.assert(runar.checkSig(sig, self.owner));
    }

    fn computeFee(amount: i64, feeBps: i64) i64 {
        return runar.percentOf(amount, feeBps);
    }

    fn scaleValue(value: i64, numerator: i64, denominator: i64) i64 {
        return runar.mulDiv(value, numerator, denominator);
    }

    fn clampValue(value: i64, lo: i64, hi: i64) i64 {
        return runar.clamp(value, lo, hi);
    }

    fn roundDown(value: i64, step: i64) i64 {
        const remainder = runar.safemod(value, step);
        return value - remainder;
    }

    pub fn deposit(self: *FunctionPatterns, sig: runar.Sig, amount: i64) void {
        self.requireOwner(sig);
        runar.assert(amount > 0);
        self.balance = self.balance + amount;
    }

    pub fn withdraw(self: *FunctionPatterns, sig: runar.Sig, amount: i64, feeBps: i64) void {
        self.requireOwner(sig);
        runar.assert(amount > 0);
        const fee = computeFee(amount, feeBps);
        const total = amount + fee;
        runar.assert(total <= self.balance);
        self.balance = self.balance - total;
    }

    pub fn scale(self: *FunctionPatterns, sig: runar.Sig, numerator: i64, denominator: i64) void {
        self.requireOwner(sig);
        self.balance = scaleValue(self.balance, numerator, denominator);
    }

    pub fn normalizeBalance(self: *FunctionPatterns, sig: runar.Sig, lo: i64, hi: i64, step: i64) void {
        self.requireOwner(sig);
        const clamped = clampValue(self.balance, lo, hi);
        self.balance = roundDown(clamped, step);
    }
};

File: conformance/tests/function-patterns/function-patterns.runar.go
//go:build ignore

package contract

import "runar"

type FunctionPatterns struct {
	runar.StatefulSmartContract
	Owner   runar.PubKey `runar:"readonly"`
	Balance runar.Bigint
}

func (c *FunctionPatterns) Deposit(sig runar.Sig, amount runar.Bigint) {
	c.requireOwner(sig)
	runar.Assert(amount > 0)
	c.Balance = c.Balance + amount
}

func (c *FunctionPatterns) Withdraw(sig runar.Sig, amount runar.Bigint, feeBps runar.Bigint) {
	c.requireOwner(sig)
	runar.Assert(amount > 0)
	fee := c.computeFee(amount, feeBps)
	total := amount + fee
	runar.Assert(total <= c.Balance)
	c.Balance = c.Balance - total
}

func (c *FunctionPatterns) Scale(sig runar.Sig, numerator runar.Bigint, denominator runar.Bigint) {
	c.requireOwner(sig)
	c.Balance = c.scaleValue(c.Balance, numerator, denominator)
}

func (c *FunctionPatterns) Normalize(sig runar.Sig, lo runar.Bigint, hi runar.Bigint, step runar.Bigint) {
	c.requireOwner(sig)
	clamped := c.clampValue(c.Balance, lo, hi)
	c.Balance = c.roundDown(clamped, step)
}

func (c *FunctionPatterns) requireOwner(sig runar.Sig) {
	runar.Assert(runar.CheckSig(sig, c.Owner))
}

func (c *FunctionPatterns) computeFee(amount runar.Bigint, feeBps runar.Bigint) runar.Bigint {
	return runar.PercentOf(amount, feeBps)
}

func (c *FunctionPatterns) scaleValue(value runar.Bigint, numerator runar.Bigint, denominator runar.Bigint) runar.Bigint {
	return runar.MulDiv(value, numerator, denominator)
}

func (c *FunctionPatterns) clampValue(value runar.Bigint, lo runar.Bigint, hi runar.Bigint) runar.Bigint {
	return runar.Clamp(value, lo, hi)
}

func (c *FunctionPatterns) roundDown(value runar.Bigint, step runar.Bigint) runar.Bigint {
	remainder := runar.Safemod(value, step)
	return value - remainder
}
